# Getting Started with Python
# ML DAY-01
Python Basics, variables, Python operators, Control structures, Functions, Basic I/O, File operations


**Python Variables:** Python has no command for declaring a variable. A variable is created the moment you first assign a value to it.

**Getting type:** You can get the data type of a variable with the type() function.

**Single or Double Quotes?** String variables can be declared either by using single or double quotes.

**Case sensitive:** Case-Sensitive Variable names are case-sensitive.

In [ ]:
# 1. Variable Declaration (no need for explicit declaration)
name = "Alice"        # String variable
age = 25              # Integer variable
height = 5.6          # Float variable
is_student = True     # Boolean variable

# 2. Getting the type of variables
print(type(name))      # Output: <class 'str'>
print(type(age))       # Output: <class 'int'>
print(type(height))    # Output: <class 'float'>
print(type(is_student))# Output: <class 'bool'>

# 3. Single or Double Quotes for Strings
msg1 = 'Hello'
msg2 = "World"
print(msg1 + " " + msg2)  # Output: Hello World

# 4. Case-Sensitive Variables
Name = "Alice"
name = "Bob"
print(Name)   # Output: Alice
print(name)   # Output: Bob

<class 'str'>
<class 'int'>
<class 'float'>
<class 'bool'>
Hello World
Alice
Bob


**Many Values to Multiple Variables:** Python allows you to assign values to multiple variables in one line

In [ ]:
x, y, z = 10, 20, 30
print(x)  # Output: 10
print(y)  # Output: 20
print(z)  # Output: 30

10
20
30


**One Value to Multiple Variables:** And you can assign the same value to multiple variables in one line

In [ ]:
a = b = c = 100
print(a)  # Output: 100
print(b)  # Output: 100
print(c)  # Output: 100

100
100
100


**Unpacking a collection:** When you have a list, python allows you to unpack that list in to the variables

In [ ]:
colors = ["red", "green", "blue"]
c1, c2, c3 = colors
print(c1)  # Output: red
print(c2)  # Output: green
print(c3)  # Output: blue

red
green
blue


**Python Casting:**

In [ ]:
a = "10"
b = "3.14"

# Convert string to int and float
int_a = int(a)       # 10
float_b = float(b)   # 3.14

# Convert int to float
x = 7
y = float(x)         # 7.0
print(y)
# Convert number to string
n = 100
text = str(n)        # "100"
print(text)

7.0
100


**Python Operators:**

Arithmetic Operators:

In [ ]:
a = 9
b = 2
print(a % b)
print(a // b)
print(a ** b)


1
4
81


Assignment Operators:

In [ ]:
x = 4
x += 3
x *= 2
print(x)

14


Comparision Operators:

In [ ]:
print(5 == 5)
print(3 != 4)
print(7 >= 10)

True
True
False


Logical Operators:

In [ ]:
a = 5
b = 2
print(a > 3 and b < 1)
print(a < 10 or b > 5)
print(not (a == 5))

False
True
False


**Control Structures:**

If-elif-else

In [ ]:
a = 33
b = 33
if b > a:
  print("b is greater than a")
elif a == b:
  print("a and b are equal")
else:
  print("a is greater than b")

a and b are equal


**and** keyword

In [ ]:
a = 200
b = 33
c = 500
if a > b and c > a:
  print("Both conditions are True")

Both conditions are True


**or** keyword

In [ ]:
a = 200
b = 33
c = 500
if a > b or a > c:
  print("At least one of the conditions is True")

At least one of the conditions is True


**not** keyword

In [ ]:
a = 33
b = 200
if not a > b:
  print("a is NOT greater than b")

a is NOT greater than b


In python we cannot do answer++ or answer--, so we need to do answer=answer+1, answer=answer-1

**Comparison Operators and Logical:**

In [ ]:
result= 5==5
print(result)
result = (2>10) and (10==10)
print(result)
result = not((2>10) and (10==10))
print(result)
isCitizen = False
age = 18
if (age>=18) or (isCitizen):
   print("allowed")

True
False
True
allowed


**For loops:**

**Loop over a range:**

In [ ]:
for i in range(3):
    print(i)

0
1
2


**Loop over a list:**

In [ ]:
subject=["maths","physics","chem"]
for i in subject:
    print(i)

print(subject[0])

maths
physics
chem
maths


**Loop over a dictionary:**

In [ ]:
person={"name":"Kinza", "age":"20"}
for key in person:
    print(key,person[key])
print("\n")

#Another way to loop over a dictionary
for key,value in person.items():
    print(key,value)
print("\n")

print(person["name"])

#value in keys can be overwritten
#It will only print the most recent one
person={"name":"Kinza", "age":"20","name":"Sheryan","age":"22"}
print("\n")
print(person["name"])
print(person["age"])

name Kinza
age 20


name Kinza
age 20


Kinza


Sheryan
22


**Nested Loops:**

In [ ]:
for i in range(2):
    for j in range(2):
        print(i,j)

0 0
0 1
1 0
1 1


**While Loops:**

In [ ]:
count=0
while count<=5:
     count=count+1
     print(count)

1
2
3
4
5
6


**Functions:**

In [ ]:
def greet(name,age):
  print(f"Hello {name} ur age is: {age}")

greet("Kinza",20)

def add(num1,num2):
   print(f"The sum is: {num1+num2}")

add(2,3)

Hello Kinza ur age is: 20
The sum is: 5


In [ ]:

def average(numbers):
 return sum(numbers)/len(numbers)

nums = [1,2,3,4,5]
print(f"The average is: {average(nums)}")

The average is: 3.0


**Lambda Function in Python:**
A lambda function in Python is a small anonymous function (i.e., it doesn’t need a name). It's used for short, simple operations — often used as a one-liner function for quick tasks.

lambda arguments: expression

In [ ]:
add = lambda num1,num2: num1+num2
print(add(2,3))

square = lambda x: x*x
print(square(2))

#filter(function, iterable)
nums = [1, 2, 3, 4, 5]
even = list(filter(lambda x: x % 2 == 0, nums))
print(even)

5
4
[2, 4]


**Basic I/O:**

In [ ]:
name = input("What is your name: ")
age = input()
print(name)
print(age)
print(f"Hello {name}, ur age is: {age}")
print("welcome " + name + " your age is " + str(age))
print("welcome",name,"your age is",age)

What is your name: kinza
20
kinza
20
Hello kinza, ur age is: 20
welcome kinza your age is 20
welcome kinza your age is 20


**File Operations:**

**Writing:**

In [ ]:
with open("data.txt","w") as file:
  file.write("Hello, World!")


**Reading:**

In [ ]:
with open("data.txt","r") as file:
  print(file.read())

Hello, World!


**Reading line by line:**

In [ ]:
file = open("data.txt", "r")
for line in file:
     print(line.strip())
     #strip removes /n
file.close()

Hello, World!


**Appending:**

In [ ]:
with open("data.txt", "a") as file:
 file.write("\nline 2")

In [ ]:
file = open("data.txt", "r")
for line in file:
     print(line.strip())
     #strip removes /n
file.close()

Hello, World!
line 2


**TASK-01: Student Report Card Generator**

 • Take Input from the User (name and age)

 • Take Input For 3 Subjects in a dictionary

subject:marks
 (english, math, urdu)

• Define a Function to Calculate Average of all subjects together

• Grade Assignment using if/elif/else

 Average Score Grade

 80 or above A

 60 to 79 B

 40 to 59 C

 Below 40 F

 • Bonus give using lambda function (Ask the user: Add 5 bonus marks to all subjects? (yes/no))

 • Write the Report to a File

 Name: Ali

 Age: 20

 Math: 90

 English: 83

 Urdu: 95

 Average: 89.33

 Grade: A

 • Read and Display the Report Card  

In [ ]:
name = input("What's your name? ")
print(name)
age = input("What's your age? ")
print(age)

subjects = ["English", "Maths", "Urdu"]
marks = {}

for subject in subjects:
    marks[subject] = int(input(f"Enter marks for {subject}: "))

bonus_input = input("Do you want to add 5 bonus marks to each score? (yes/no): ").strip().lower()
add_bonus = lambda x: x + 5 if bonus_input == "yes" else x

for subject in subjects:
    marks[subject] = add_bonus(marks[subject])

def avg(scores):
    return sum(scores) / len(scores)

average = avg(list(marks.values()))
print(f"Average Marks: {average:.2f}")

if average >= 80:
    grade = "A"
elif 60 <= average <= 79:
    grade = "B"
elif 40 <= average <= 59:
    grade = "C"
else:
    grade = "F"

print(f"Grade: {grade}")
with open("studentreport.txt", "w") as file:
    file.write(f"Name: {name}\n")
    file.write(f"Age: {age}\n")
    for subject in subjects:
        file.write(f"{subject}: {marks[subject]}\n")
    file.write(f"Average: {average:.2f}\n")
    file.write(f"Grade: {grade}\n")

print("Student data has been written to 'studentreport.txt'.")

with open("studentreport.txt", "r") as file:
    print("\nGenerated Report Card:")
    print(file.read())

What's your name? Kinza
Kinza
What's your age? 20
20
Enter marks for English: 83
Enter marks for Maths: 90
Enter marks for Urdu: 95
Do you want to add 5 bonus marks to each score? (yes/no): yes
Average Marks: 94.33
Grade: A
Student data has been written to 'studentreport.txt'.

Generated Report Card:
Name: Kinza
Age: 20
English: 88
Maths: 95
Urdu: 100
Average: 94.33
Grade: A



# ML DAY-02

Lists, Dictionaries, Sets

**Lists:**

List items are ordered, changeable, and allow duplicate values.

List items are indexed, the first item has index [0], the second item has index [1] etc.

The list is changeable, meaning that we can change, add, and remove items in a list after it has been created.

Lists can have elements of multiple data types.

Made by square brackets. []

In [ ]:
numbers = [1,2,3,4,5]
print(numbers)

mixed = ["Kinza",1,2,2.55,True]
print(mixed)

[1, 2, 3, 4, 5]
['Kinza', 1, 2, 2.55, True]


**Length:** use len(listname)

**Type:** use type(listname)

In [ ]:
thislist = ["apple", "banana", "cherry"]
print(len(thislist))

mylist = ["apple", "banana", "cherry"]
print(type(mylist))

3
<class 'list'>


**List Methods:**
append, remove, insert, pop

1. append(value)
What it does: Adds a value to the end of the list

2. insert(index, value)
What it does: Inserts a value into the list at a specific position (index).
All elements from that position onward are shifted one place to the right.

3. remove(value)
What it does: Removes the first occurrence of the specified value from the list.

Important: If the value appears more than once, only the first one is removed.

4. pop()
What it does: Removes and returns the last item from the list.

Useful when: You want to both remove the item and use it elsewhere in your code.

In [ ]:
mixed = ["Kinza",1,2,2.55,True]
print(mixed)
mixed.append("Sheryan")
print(mixed)
#this adds 2 to the list at index 0
mixed.insert(0,2)
print(mixed)

mixed.remove(2.55)
print(mixed)

popped=mixed.pop()
print(popped)
print(mixed)

mixed.append("Sheryan")
print(mixed)

['Kinza', 1, 2, 2.55, True]
['Kinza', 1, 2, 2.55, True, 'Sheryan']
[2, 'Kinza', 1, 2, 2.55, True, 'Sheryan']
[2, 'Kinza', 1, 2, True, 'Sheryan']
Sheryan
[2, 'Kinza', 1, 2, True]
[2, 'Kinza', 1, 2, True, 'Sheryan']


**List Comprehensions:**
List comprehension is a short and elegant way to create a new list by performing an operation on each item in an existing list (or any iterable like a string, range, etc.).

[expression for item in iterable]

In [ ]:
squares =[x**2 for x in range(5)]
print(squares)

b=[]
for i in range(5):
    b.append(i**2)
print(b)

[0, 1, 4, 9, 16]
[0, 1, 4, 9, 16]


**Nested list comprehensions:**

[[expression for item in outer] for outer in iterable]

In [ ]:

#Multiplication from 1 to 5
table = [[i * j for j in range(1, 6)] for i in range(1, 6)]
for row in table:
    print(row)

print(table)

[1, 2, 3, 4, 5]
[2, 4, 6, 8, 10]
[3, 6, 9, 12, 15]
[4, 8, 12, 16, 20]
[5, 10, 15, 20, 25]
[[1, 2, 3, 4, 5], [2, 4, 6, 8, 10], [3, 6, 9, 12, 15], [4, 8, 12, 16, 20], [5, 10, 15, 20, 25]]


**Unpacking elements with *:**

In [ ]:

#This will give error
#a,b,c=[1,2,3,4,5]
#print(a)
#print(b)
#print(c)

#This will assign 1 to a, 5 to c and rest 2,3,4 to b
a,*b,c=[1,2,3,4,5]
print(a)
print(b)
print(c)
print("\n")

#This will assign 1 to a, 2 to b and 3,4,5 to c
a,b,*c=[1,2,3,4,5]
print(a)
print(b)
print(c)

1
[2, 3, 4]
5


1
2
[3, 4, 5]


**Dictionary:**
A dictionary in Python is an unordered, mutable collection of key-value pairs. It's like a real-life dictionary where each word (key) maps to a definition (value).

Made with {} curly brackets.

In [ ]:
student={
   "name":"Kinza",
    "age":20,
    "grade":"A"
}
print(student["name"])
print(student.get("age"))

for key in student:
    print(student[key])

for key,value in student.items():
    print(key,value)

print(student.keys())
print(student.values())
print(student.items())

#Adding a key
student["good teacher"] ="Sir Moheez"
print(student)

print("\n")
# key can also be like index
cars={
  1:1,
  2:2,
  3:3
}
print(cars)

Kinza
20
Kinza
20
A
name Kinza
age 20
grade A
dict_keys(['name', 'age', 'grade'])
dict_values(['Kinza', 20, 'A'])
dict_items([('name', 'Kinza'), ('age', 20), ('grade', 'A')])
{'name': 'Kinza', 'age': 20, 'grade': 'A', 'good teacher': 'Sir Moheez'}


{1: 1, 2: 2, 3: 3}


In [ ]:
squares={
    x:x**2 for x in range(5)
}
print(squares)

{0: 0, 1: 1, 2: 4, 3: 9, 4: 16}


**Difference between list and dictionary:**


A **list** is a collection of items arranged in a specific order. Each item in a list is accessed by its index, starting from 0. Lists can hold any type of data—numbers, strings, or even other lists—and they allow duplicate values. You use lists when the position of elements matters or when you just want to store a sequence of items.

Example: A list of student marks like [85, 90, 78].


A **dictionary**, on the other hand, is a collection of key-value pairs. Each item is stored as a pair where the key is unique and is used to access its corresponding value. Unlike lists, dictionaries are unordered (until Python 3.7+, where insertion order is preserved), and you access values using the key, not an index. Dictionaries are great when you want to label data clearly.

Example: A student record like {"name": "Ali", "age": 20, "grade": "A"}.

**Sets:**

A set in Python is a collection of unique, unordered elements. It is similar to a mathematical set — it automatically removes duplicates and doesn't care about the order of elements.

**Key features of set:**

Unique

Unordered

Mutable

Heterogeneous (Can contain different Datatypes)

In [ ]:
colours={'red','green','black'}
print(colours)

#Set cannot have duplicate elements, ek dafa hi print krega
#so we say set is a collection of unique elements
colours={'red','green','black','red'}
print(colours)

#we can also create set from a list
colours=set(['red','green','black'])
print(colours)

colours=set(['red','green','black','red'])
print(colours)

num1={1,2,3,4}
num2={3,4,5,6}

#Union
union=num1 | num2
print(union)

#Intersectiom
intersection= num1 & num2
print(intersection)

#Difference
diff= num1-num2
print(diff)

#Common Set Operations:
num1.add(5)
print(num1)

#Remove gives error if element not found
num1.remove(1)
print(num1)

#Discard (does nothing if element not found)
num1.discard(1)
print(num1)

{'red', 'green', 'black'}
{'red', 'green', 'black'}
{'red', 'green', 'black'}
{'red', 'green', 'black'}
{1, 2, 3, 4, 5, 6}
{3, 4}
{1, 2}
{1, 2, 3, 4, 5}
{2, 3, 4, 5}
{2, 3, 4, 5}


**String Operations:**

Changing case

In [ ]:
s = "Hello World"
print(s.lower())     # hello world
print(s.upper())     # HELLO WORLD
print(s.title())     # Hello World
print(s.capitalize())  # Hello world

hello world
HELLO WORLD
Hello World
Hello world


Removing Whitespaces

In [ ]:
s = "  hello  "
print(s.strip())   # "hello"
print(s.lstrip())  # "hello  "
print(s.rstrip())  # "  hello"

hello
hello  
  hello


Replacing Substrings

In [ ]:
s = "I love Java"
print(s.replace("Java", "Python"))

I love Python


Splitting and Joining Strings

In [ ]:
s = "apple,banana,grape"
fruits = s.split(",")  # ['apple', 'banana', 'grape']
print(fruits)

sentence = " ".join(fruits)
print(sentence)  # apple banana grape

['apple', 'banana', 'grape']
apple banana grape


Finding substrings

In [ ]:
s = "data science"
print(s.find("sci"))     # 5
print(s.find("xyz"))     # -1
print("sci" in s)        # True

5
-1
True


String Indexing & Slicing

In [ ]:
s = "Python"
print(s[0])     # P
print(s[-1])    # n
print(s[1:4])   # yth
print(s[::-1])  # nohtyP (reverse)

P
n
yth
nohtyP


Checking String Content:

In [ ]:
s = "123"
print(s.isdigit())   # True

s = "abc"
print(s.isalpha())   # True

s = "abc123"
print(s.isalnum())   # True

s = "   "
print(s.isspace())   # True

True
True
True
True


Reversing a String

In [ ]:
s = "hello"
print(s[::-1])  # olleh

Counting Occurences

In [ ]:
s="appleia"
print(s.count("a"))

2


# ML DAY-03
Exception Handling, OOP and Numpy Operations

**Some Common Built-in Exceptions:**

ValueError:Raised when a function receive an argumentof correct type but inappropriate value.

ZeroDivisionError: Raised when division or modulo by zero occurs.

TypeError: Raised when an operation is applied to an object of inappropriate type.

IndexError: Raised when a sequence index is out of range.

KeyError: Raised when a dictionary key is not found.

NameError: Raised when a variable is not defined.

OverflowError: Raised when a number is too large to be represented.

RuntimeError: Raised when an error is detected that doesn’t fall in any category.

In [ ]:
try:
   number =int(input("Enter a number"))
   print(number)
   answer=10/number
   print(answer)
except ZeroDivisionError:
   print("Cannot divide by zero")
except ValueError:
   print("Invalid input!")
except Exception as e:
   print(f"An error occured {e}")
else:
   print("No exception occured")
finally:
   print("This line always runs")

Enter a number0
0
Cannot divide by zero
This line always runs


**Custom Error:**

Made by user

In [ ]:
class MyCustomError(Exception):
     pass

def check_age(age):
    if age<0:
        raise MyCustomError("Age cannot be negative")

try:
    age = int(input("Enter your age: "))
    check_age(age)
    print("Your age is:", age)
except MyCustomError as e:
    print("Custom Error:", e)

Enter your age: -1
Custom Error: Age cannot be negative


**OOP:**

OOP (Object-Oriented Programming) is a programming paradigm based on the concept of "objects", which can contain data (attributes) and code (methods).

**🔹 What is a Decorator?**
A decorator is a function that takes another function as input and extends or modifies its behavior without changing its source code.

In [ ]:
class Car:
    # This acts as a constructor
    def __init__(self, carname, model, price):
        self.carname = carname
        self.model = model
        self.price = price

    def print(self):
        print(f"Car Name: {self.carname}")
        print(f"Model: {self.model}")
        print(f"Price: {self.price}")

porsche = Car("Porsche", "911", 1000000)
porsche.print()

Car Name: Porsche
Model: 911
Price: 1000000


In [ ]:
class Car:
    # This acts as a constructor
    def __init__(self, carname, model, price):
        self.carname = carname
        self.model = model
        self.price = price

    @property
    def print(self):
        print(f"Car Name: {self.carname}")
        print(f"Model: {self.model}")
        print(f"Price: {self.price}")

porsche = Car("Porsche", "911", 1000000)
porsche.print

#Decorator ki waja se we can call print function as variable

Car Name: Porsche
Model: 911
Price: 1000000


In [ ]:
class Student:
    def __init__(self, name, age):
        self.name = name
        self.__age = age  # private attribute

    def get_age(self):        # public method
        return self.__age

    def set_age(self, age):
        if age > 0:
            self.__age = age

stu = Student("Kinza",20)
print(stu.get_age())
stu.set_age(21)
print(stu.get_age())

20
21


In [ ]:
class Test:
    def __init__(self):
        print("Constructor called")

    def __del__(self):
        print("Destructor called")

obj = Test()

Constructor called


In [ ]:
class Parent:
    def speak(self):
        print("I'm parent class")

class Child(Parent):
    def speakk(self):
        print("I'm child class")

c= Child()
c.speakk()
c.speak()

I'm child class
I'm parent class


In [ ]:
class Parent:
    def __init__(self, parent_name):
        self.parent_name = parent_name

    def speak(self):
        return f"I'm the parent: {self.parent_name}"

class Child(Parent):
    def __init__(self, parent_name, child_name):
        super().__init__(parent_name)  # Call Parent's constructor
        self.child_name = child_name

    def speakk(self):
        return f"I'm the child: {self.child_name}"

# Create object
c = Child("Sara", "Kinza")
print(c.speakk())     # I'm the child: Kinza
print(c.speak())      # I'm the parent: sara

I'm the child: Kinza
I'm the parent: Sara


**NumPy (Numerical Python)** is a library used for:

Fast numerical operations

Multidimensional arrays (ndarrays)

Matrix operations

Broadcasting

Linear algebra, statistics, and more

In [ ]:
import numpy as np
a = np.array([1, 2, 3])
b = np.array([[1, 2], [3, 4]])

print(a)         # 1D array
print(b)         # 2D array
print(a.shape)   # (3,)
print(b.shape)   # (2, 2)

[1 2 3]
[[1 2]
 [3 4]]
(3,)
(2, 2)


In [ ]:
arr = np.array([10, 20, 30, 40, 50])
print(arr[0])      # 10
print(arr[-1])     # 50
print(arr[1:4])    # [20 30 40]
#This includes 1st index value and excludes 4th index value

10
50
[20 30 40]


In [ ]:
arr=np.array([[1,2,3],[4,5,6]])
print(arr)

# Reshaping an array
reshaped = arr.reshape(3,2)
print(reshaped)

# Transpose of array
print(arr.T)

[[1 2 3]
 [4 5 6]]
[[1 2]
 [3 4]
 [5 6]]
[[1 4]
 [2 5]
 [3 6]]


In [ ]:
arr=np.array([[1,2,3],[4,5,6]])
print(arr)

print(np.sum(arr))
print(np.mean(arr))
print(np.min(arr))
print(np.max(arr))

[[1 2 3]
 [4 5 6]]
21
3.5
1
6


In [ ]:
a = np.array([1, 2, 3])
b = 2
print(a * b)     # [2 4 6]

[2 4 6]


In [ ]:
import numpy as np
zeroes =np.zeros((3,2))
print(zeroes)

ones = np.ones((3,2))
print(ones)

identity = np.eye((2))
print(identity)

random = np.random.rand(2,3)
print(random)

# 3 values in range from 2 to 10
random = np.random.randint(2,10,3)
print(random)

fill = np.full((2,2),2)
print(fill)

fill = np.full((2,2),10)
print(fill)

# np.arange(start,stop,step)
# 1 is included 10 is excluded
arr = np.arange(1,10,2)
print(arr)

# 0 is included 10 is excluded
arr = np.arange(0,10,2)
print(arr)

[[0. 0.]
 [0. 0.]
 [0. 0.]]
[[1. 1.]
 [1. 1.]
 [1. 1.]]
[[1. 0.]
 [0. 1.]]
[[0.09096783 0.48180266 0.81849187]
 [0.52234478 0.63929921 0.63834194]]
[6 3 5]
[[2 2]
 [2 2]]
[[10 10]
 [10 10]]
[1 3 5 7 9]
[0 2 4 6 8]


**arr.shape:**

Returns the shape of array. (rows, columns)


**arr.size:**

Returns the total number of elements in the array

**arr.dtype:**

Shows the data type of the elements (e.g., int32, int64, float64).


**arr.ndim:**

Returns the number of dimensions (axes).

📌 A 1D array has ndim = 1.



In [ ]:
arr = np.array([1,2,3,4,5,6])
# To know the shape
print(arr.shape)
# To know the size
print(arr.size)
# To know the type
print(arr.dtype)
# To know the dimension
print(arr.ndim)

arr_1d = np.array([1,2,3,4,5,6])
arr_2d = np.array([[1,2,3], [4,5,6]])
arr_3d = np.array([[[1,2,3], [4,5,6]]])

print(arr_1d.ndim)
print(arr_2d.ndim)
print(arr_3d.ndim)

(6,)
6
int64
1
1
2
3


**astype()** is used to convert the data type of an array without modifying the original array. This is helpful in many scenarios:

In [ ]:
arr = np.array([1.2,2.4,3.6])
print(arr.dtype)

print(arr.astype(int))
print(arr.dtype)

float64
[1 2 3]
float64


# ML DAY -04
Advanced Numpy Operations

In [ ]:
import numpy as np
arr = np.array([1,2,3,4,5,6])
print(arr+5)
print(arr*2)
print(arr**2)

[ 6  7  8  9 10 11]
[ 2  4  6  8 10 12]
[ 1  4  9 16 25 36]


In [ ]:
import numpy as np
numpy_array1 = np.array([1,2,3])
numpy_array2 = np.array([4,5,6])
numpy_array3 = numpy_array1+numpy_array2
print(numpy_array3)

numpy_array4 = numpy_array1-numpy_array2
print(numpy_array4)

numpy_array5 = numpy_array1*numpy_array2
print(numpy_array5)

# Dot product (1*4 + 2*5 + 3*6 = 32)
# same as np.dot()
numpy_array6 = numpy_array1@numpy_array2
print(numpy_array6)

# [1 // 4, 2 // 5, 3 // 6] → [0, 0, 0]
numpy_array7 = numpy_array1 // numpy_array2
print(numpy_array7)

#[1**4, 2**5, 3**6] → [1, 32, 729]
numpy_array8 = numpy_array1 ** numpy_array2
print(numpy_array8)

[5 7 9]
[-3 -3 -3]
[ 4 10 18]
32
[0 0 0]
[  1  32 729]


**Aggregate Functions:**

np.sum(arr)

np.mean(arr) --> For average of array

np.min(arr) --> For minimum of array

np.max(arr) --> For maximum of array

np.std(arr) --> For standard deviation of array

np.var(arr) --> For variance of array

np.sum(arr,axis=0) --> Column-wise sum

np.sum(arr,axis=1) --> Row-wise sum

In [ ]:
import numpy as np
arr = np.array([[1,2,3],[4,5,6]])
print(arr)
print(np.sum(arr))
print(np.mean(arr))
print(np.min(arr))
print(np.max(arr))
print(np.std(arr))
print(np.var(arr))
# Column- wise sum
print(np.sum(arr,axis=0))
# Row- wise sum
print(np.sum(arr,axis=1))

[[1 2 3]
 [4 5 6]]
21
3.5
1
6
1.707825127659933
2.9166666666666665
[5 7 9]
[ 6 15]


**Indexing and Slicing:**

 Slicing Syntax: arr[start:stop:step]

 📌 Extra Tip: Slicing is non-destructive
It returns a view or copy of the data — the original array remains unchanged unless you assign to the slice.

In [ ]:
arr = np.array([10,20,30,40,50])
print(arr[0]) # first element
print(arr[2]) # third elemnt
print(arr[-1]) # last element

# arr[start:end]
print(arr[1:4]) # 1 index is included, 4th index is excluded
print(arr[:3]) # starting from 0th index, ending at 3rd index is excluded
print(arr[2:]) # starting from 2nd index
print(arr[::2]) # picks every 2nd element from start to end
print(arr[::-1]) # Reversing the whole array

10
30
50
[20 30 40]
[10 20 30]
[30 40 50]
[10 30 50]
[50 40 30 20 10]


**Fancy Indexing:**
Fancy indexing allows you to access multiple array elements at once using arrays (or lists) of indices, rather than using slicing.

In [ ]:
import numpy as np

arr = np.array([10, 20, 30, 40, 50])
indices = [0, 2, 4]

print(arr[indices])  # Output: [10 30 50]
# or
print(arr[[0,2,4]])

[10 30 50]
[10 30 50]


**Filtering:**
Filtering is a technique to extract elements from an array that satisfy a specific condition. It uses Boolean indexing under the hood.



In [ ]:
arr = np.array([10,20,30,40,50])
print(arr[arr>25])
# it will return only those values from arr that are greater than 25

[30 40 50]


**Re-shaping:**
Reshaping is changing the shape (rows, columns, dimensions) of a NumPy array without changing its data.

It’s useful in:

Preprocessing data for ML models

Converting between 1D, 2D, and 3D structures

Making arrays compatible for operations


Syntax: reshape(rows,columns) specify new shape
        if dimensions match

Reshaping doesnot create a copy, it just returns the view.

In [ ]:
arr = np.array([1, 2, 3, 4, 5, 6])
reshaped = arr.reshape(2, 3)

print(reshaped)

[[1 2 3]
 [4 5 6]]


**Flattening array:**
converting multidimensional array to 1-d array.

.ravel() -> views, original array affected.

.flatten() -> returns a copy, original array not affected.


In [ ]:
arr_2d = np.array([[1,2,3],[4,5,6]])
print(arr_2d)
print(arr_2d.ravel())
print(arr_2d.flatten())

arr3d = np.array([[[1, 2], [3, 4]], [[5, 6], [7, 8]]])
print(arr3d.flatten())     # Output: [1 2 3 4 5 6 7 8]

[[1 2 3]
 [4 5 6]]
[1 2 3 4 5 6]
[1 2 3 4 5 6]
[1 2 3 4 5 6 7 8]


**Array Modification:**

np.insert(arr, index, value, axis=None)

np.append(arr, values, axis=None)

np.delete(arr, obj, axis=None) where obj is Index/indices to delete (can be int, list, or slice)

In [ ]:
arr = np.array([1, 2, 5])
new_arr = np.insert(arr, [2], [3, 4])
# inserts 3,4 at index 2
print(new_arr)  # Output: [1 2 3 4 5]

arr = np.array([10, 20, 30])
new_arr = np.insert(arr, 1, 15)

print(new_arr)  # Output: [10 15 20 30]

[1 2 3 4 5]
[10 15 20 30]


In [ ]:
arr2d = np.array([[1, 2], [3, 4]])

# Insert a new row at index 1
new_arr = np.insert(arr2d, 1, [9, 9], axis=0)

print(new_arr)

[[1 2]
 [9 9]
 [3 4]]


In [ ]:
arr = np.array([10,20,30,40,50,60])
print(arr)
new_arr = np.append(arr,[70,80])
print(new_arr)

arr = np.array([[10,20,30],[40,50,60]])
print(arr)
new_arr=np.append(arr,[[70,80,90]],axis=0)
print(new_arr)

new_arr=np.append(arr,[[70],[80]],axis=1)
print(new_arr)

[10 20 30 40 50 60]
[10 20 30 40 50 60 70 80]
[[10 20 30]
 [40 50 60]]
[[10 20 30]
 [40 50 60]
 [70 80 90]]
[[10 20 30 70]
 [40 50 60 80]]


In [ ]:
arr = np.array([10, 20, 30, 40, 50])
new_arr = np.delete(arr, 2)
#deletes the value at index 2
print(new_arr)

[10 20 40 50]


In [ ]:
arr_2d = np.array([[1,2,3],[4,5,6]])
print(arr_2d)
#0th row will be deleted
print(np.delete(arr_2d,0,axis=0))
#0th column will be deleted
print(np.delete(arr_2d,0,axis=1))

[[1 2 3]
 [4 5 6]]
[[4 5 6]]
[[2 3]
 [5 6]]


np.concatenate([arr1,arr2])

or

np.concatenate((arr1,arr2),axis)

axis = 0 --> Vertical stacking (row-wise)

axis = 1 --> Horizontal stacking (column-wise)

In [ ]:
import numpy as np
arr1 = np.array([1,2,3])
arr2 = np.array([4,5,6])
arr3 = np.concatenate([arr1,arr2])
print(arr3)

arr3 = np.concatenate((arr1,arr2))
print(arr3)

[1 2 3 4 5 6]
[1 2 3 4 5 6]


# ML DAY- 05
Numpy Advanced

vstack() -->Vertically stack (row-wise)

hstack() --> Horizontally stack (column-wise)

In [ ]:
import numpy as np
arr1 = np.array([1,2,3])
arr2 = np.array([4,5,6])
print(np.vstack((arr1,arr2)))
print(np.hstack((arr1,arr2)))

[[1 2 3]
 [4 5 6]]
[1 2 3 4 5 6]


**Splitting Array:**

np.split() --> equal parts mai split

np.vsplit() --> vertically split

np.hsplit() --> horizontally split

In [ ]:
arr = np.array([1,2,3,4,5,6])
print(np.split(arr,3)) #This will split the error into 3 equal parts

# arr = np.array([1, 2, 3, 4, 5])
# np.split(arr, 3)  # ❌ This will raise an error

[array([1, 2]), array([3, 4]), array([5, 6])]


**Broadcasting:**
Broadcasting allows NumPy to perform operations on arrays of different shapes as if they were the same shape.

It automatically expands the smaller array without copying data, allowing fast computation.



In [ ]:
prices = [100,200,300,400]
discount = 10
final_prices=[]
for x in prices:
  final_prices.append(x-(x*discount/100))
print(final_prices)

# Instead of loops we can use broadcasting
# Broadcasting --> fast, no need of loops
prices = np.array([100,200,300,400])
discount=10
final_prices = prices-(prices*discount/100)
print(final_prices)

[90.0, 180.0, 270.0, 360.0]
[ 90. 180. 270. 360.]


How numpy handles arrays of different shapes?

1) Matching dimension --> Both arrays of same size.

2) Expanding: When the shapes differ, NumPy tries to automatically expand (broadcast) the smaller array without copying data, using these broadcasting rules:

✅ Broadcasting Rules:
Compare dimensions from the rightmost.

Two dimensions are compatible if:

They are equal, or

One of them is 1

NumPy stretches dimensions of size 1 to match the other.

3) Numpy throws error if incompatible shapes.

In [ ]:
arr = np.array([100,200,300,400,500,600])
result = arr*2
print(result)

[ 200  400  600  800 1000 1200]


In [ ]:
# Broadcasting 1d array to 2d array
matrix = np.array([[1,2,3],[4,5,6]])
vector = np.array([10,20,30])
result = matrix+vector
print(result)

[[11 22 33]
 [14 25 36]]


In [ ]:
# Incompatible shapes --> Error
arr1 = np.array([[1,2,3],[4,5,6]]) #shape-->(2,3)
arr2 = np.array([10,20])          #shape-->(2,)
print(arr+arr2)
# will throw error

**Comparison:**

In [ ]:
# it basically performs if functionality but more readable
arr1 = np.array([1,2,3])
arr2 = np.array([5,2,9])
arr3 = arr1 == arr2
print(arr3)
arr3 = np.array((1,2,3))
arr4 = (arr1 > arr2) & arr2 == arr3 # multiple conditions
print(arr4)

#copies only when arr 1 is greater than arr2 else dropped
arr4 = arr2[arr2>=5]
print(arr4)# 2 dropped

arr4[arr4==5]= 10
print(arr4)# replaced 5 with 10

[False  True False]
[False False False]
[5 9]
[10  9]


In [ ]:
arr1 = np.array(['True','False','True'])
arr2 = np.array(['True','False','True'])
arr5 = np.logical_and(arr1,arr2)
print(arr5)
# broadcasting rule is, either size should be equal
# else any one of them should be 1 (row/col)

[ True  True  True]


**Basic Arthematic operations**

In [ ]:
arr1 = np.array([1, 2, 4])
arr2 = np.array([6, 7, 8])
arr3=np.add(arr1,arr2)
print(arr3)

arr4=np.subtract(arr1,arr2)
print(arr4)

arr5 = np.multiply(arr1,arr2)
print(arr5)

arr6 = np.divide(arr1,arr2)
print(arr6)

arr7 = np.sqrt(arr1)
print(arr7)

arr8 = np.log(arr1)
print(arr8)

[ 7  9 12]
[-5 -5 -4]
[ 6 14 32]
[0.16666667 0.28571429 0.5       ]
[1.         1.41421356 2.        ]
[0.         0.69314718 1.38629436]


In [ ]:
matrix_a = np.array([1, 2, 3, 4, 5])
matrix_b = np.array([6, 7, 8, 9, 10])

print(matrix_a)
print(matrix_b)

matrix_c = np.add(matrix_a, matrix_b)
print(matrix_c)

matrix_d = np.subtract(matrix_a, matrix_b)
print(matrix_d)

matrix_e = np.multiply(matrix_a, matrix_b)
print(matrix_e)

matrix_f = np.divide(matrix_a, matrix_b)
print(matrix_f)

dot_product = np.dot(matrix_a, matrix_b)
print(dot_product)

transpose_a = np.transpose(matrix_a)
print(transpose_a)

transpose_b = np.transpose(matrix_b)
print(transpose_b)

[1 2 3 4 5]
[ 6  7  8  9 10]
[ 7  9 11 13 15]
[-5 -5 -5 -5 -5]
[ 6 14 24 36 50]
[0.16666667 0.28571429 0.375      0.44444444 0.5       ]
130
[1 2 3 4 5]
[ 6  7  8  9 10]


**Statistics:**
Percentiles, Quartiles, IQR

25th percentile (Q1): This is the value below which 25% of the data fall.

Between 1st and 2nd element → 20.0

50th percentile (Median / Q2): Middle value of the sorted data.

Middle value → 30.0

75th percentile (Q3): Value below which 75% of the data fall.

Between 4th and 5th element → 40.0

IQR = Q3 - Q1

Q1 = 25th percentile

Q3 = 75th percentile

**Outliers Detection:**

< Q1 - 1.5 * IQR   or   > Q3 + 1.5 * IQR

In [ ]:
arr1 = np.array([10,20,30,40,50])
percentile1=np.percentile(arr1,25)
print(percentile1)

percentile2=np.percentile(arr1,50)
print(percentile2)

percentile3=np.percentile(arr1,75)
print(percentile3)

# IQR = Q3-Q1
IQR = percentile3-percentile1
print(IQR)

stddev = np.std(arr1)
print(stddev)

variance = np.var(arr1)
print(variance)

minimum = np.min(arr1)
print(minimum)

maximum = np.max(arr1)
print(maximum)

mean = np.mean(arr1)
print(mean)

median = np.median(arr1)
print(median)

# For outliers
lower_bound = percentile1- 1.5*IQR
upper_bound = percentile3 + 1.5*IQR
print(lower_bound)
print(upper_bound)

outliers=arr1 [(arr1<lower_bound) | (arr1>upper_bound)]
print(outliers)

20.0
30.0
40.0
20.0
14.142135623730951
200.0
10
50
30.0
30.0
-10.0
70.0
[]


**Handling Missing Values:**

np.isnan --> detects missing values
returns true for nan values

np.nan_to_num()	--> Replace NaNs with zero (or custom)

In [ ]:
arr = np.array([10, 20, np.nan, 40, np.nan, 60])

print("Original array:", arr)
print("Where values are NaN:", np.isnan(arr))

arr_filled = np.nan_to_num(arr)
print("Array with NaN replaced by 0:", arr_filled)

# Replaces nan with 100
arr_filled = np.nan_to_num(arr,nan=100)
print("Array with NaN replaced by 0:", arr_filled)

Original array: [10. 20. nan 40. nan 60.]
Where values are NaN: [False False  True False  True False]
Array with NaN replaced by 0: [10. 20.  0. 40.  0. 60.]
Array with NaN replaced by 0: [ 10.  20. 100.  40. 100.  60.]


**Tasks Day -04**

**Mini Project:**
You are given daily sales data (in USD) of 6 products over 10 days. Perform the following tasks.

**Q1. Mathematical Operations**   

a. Calculate the total sales per product over 10 days.

b. Calculate the average daily sales per product.

c. Increase each sale by a 5% commission using broadcasting.

d. Apply the square root to all sales values (for testing purposes).

**Q2. Broadcasting Concepts**

a. Create a 1D and using broadcasting.

bonus_array = np.array([10, 20, 15, 25, 30, 5]) add it to each day’s sales

b. Add a $50 flat bonus to each sale using broadcasting.

**Q3. Statistical Analysis**

a. Find the mean, median, variance, and standard deviation of the entire dataset.

b. Find the maximum and minimum sale value and calculate the range.

c. Calculate the interquartile range (IQR) of all values.

**Q4. Logical & Comparison Operations**

a. Identify and list all sales greater than $250.

b. Replace all values greater than $300 with 300 (cap the max sale).

c. Count how many times sales were between $200 and  $250.

d. Create a new array showing only the sales below the mean.

**Q5. Searching, Sorting, and Final Summary**

a. Sort the sales of Day 5 in ascending order.

b. Find the day (row index) with the highest total sales.

c. Calculate column-wise means (average per product).

d. Calculate row-wise means (average per day).

e. Print the overall average sales.

**Q6. Bonus**

Create a function that returns all values 2 standard deviations above and below
the mean.

In [ ]:
import numpy as np

sales_data = np.array([
    [200, 220, 250, 210, 180, 190],  # Day 1
    [230, 240, 260, 200, 195, 205],  # Day 2
    [210, 215, 255, 220, 185, 200],  # Day 3
    [205, 225, 275, 190, 195, 195],  # Day 4
    [215, 230, 265, 225, 200, 210],  # Day 5
    [225, 235, 275, 230, 205, 215],  # Day 6
    [235, 245, 280, 240, 210, 220],  # Day 7
    [245, 255, 290, 205, 215, 225],  # Day 8
    [255, 265, 300, 260, 220, 230],  # Day 9
    [265, 275, 310, 270, 225, 235]   # Day 10
])
print(sales_data)

#total sales per product over 10 days
# so we need column wise sum

print(np.sum(sales_data,axis =0))
print(np.mean(sales_data,axis=0))

final_sales = sales_data + (sales_data*5/100)
print(final_sales)

print(np.sqrt(sales_data))

[[200 220 250 210 180 190]
 [230 240 260 200 195 205]
 [210 215 255 220 185 200]
 [205 225 275 190 195 195]
 [215 230 265 225 200 210]
 [225 235 275 230 205 215]
 [235 245 280 240 210 220]
 [245 255 290 205 215 225]
 [255 265 300 260 220 230]
 [265 275 310 270 225 235]]
[2285 2405 2760 2250 2030 2125]
[228.5 240.5 276.  225.  203.  212.5]
[[210.   231.   262.5  220.5  189.   199.5 ]
 [241.5  252.   273.   210.   204.75 215.25]
 [220.5  225.75 267.75 231.   194.25 210.  ]
 [215.25 236.25 288.75 199.5  204.75 204.75]
 [225.75 241.5  278.25 236.25 210.   220.5 ]
 [236.25 246.75 288.75 241.5  215.25 225.75]
 [246.75 257.25 294.   252.   220.5  231.  ]
 [257.25 267.75 304.5  215.25 225.75 236.25]
 [267.75 278.25 315.   273.   231.   241.5 ]
 [278.25 288.75 325.5  283.5  236.25 246.75]]
[[14.14213562 14.83239697 15.8113883  14.49137675 13.41640786 13.78404875]
 [15.16575089 15.49193338 16.1245155  14.14213562 13.96424004 14.31782106]
 [14.49137675 14.6628783  15.96871942 14.83239697 13.60147

In [ ]:
bonus_array = np.array([10, 20, 15, 25, 30, 5])
print(bonus_array)
print(bonus_array+sales_data)

print(sales_data+50)

[10 20 15 25 30  5]
[[210 240 265 235 210 195]
 [240 260 275 225 225 210]
 [220 235 270 245 215 205]
 [215 245 290 215 225 200]
 [225 250 280 250 230 215]
 [235 255 290 255 235 220]
 [245 265 295 265 240 225]
 [255 275 305 230 245 230]
 [265 285 315 285 250 235]
 [275 295 325 295 255 240]]
[[250 270 300 260 230 240]
 [280 290 310 250 245 255]
 [260 265 305 270 235 250]
 [255 275 325 240 245 245]
 [265 280 315 275 250 260]
 [275 285 325 280 255 265]
 [285 295 330 290 260 270]
 [295 305 340 255 265 275]
 [305 315 350 310 270 280]
 [315 325 360 320 275 285]]


In [ ]:
mean = np.mean(sales_data)
Stddev = np.std(sales_data)
Variance = np.var(sales_data)
median = np.median(sales_data)
print(mean)
print(median)
print(Stddev)
print(Variance)

Max = np.max(sales_data)
Min = np.min(sales_data)
Range = Max - Min
print(Max)
print(Min)
print(Range)

Q1 = np.percentile(sales_data,25)
Q2 = np.percentile(sales_data,50)
Q3 = np.percentile(sales_data,75)
IQR = Q3-Q1
print(Q1)
print(Q2)
print(Q3)
print(IQR)

230.91666666666666
225.0
30.006827926693987
900.4097222222222
310
180
130
208.75
225.0
255.0
46.25


In [ ]:
greater_than_250 = sales_data[(sales_data>250)]
print(greater_than_250)

capped_sales = np.copy(sales_data)
capped_sales[capped_sales > 300] = 300
print(capped_sales)


between_200and250 = sales_data[(sales_data>200) & (sales_data<250)]
print(between_200and250)
print(np.size(between_200and250))

new_arr = sales_data[(sales_data<np.mean(sales_data))]
print(new_arr)

[260 255 275 265 275 280 255 290 255 265 300 260 265 275 300 270]
[[200 220 250 210 180 190]
 [230 240 260 200 195 205]
 [210 215 255 220 185 200]
 [205 225 275 190 195 195]
 [215 230 265 225 200 210]
 [225 235 275 230 205 215]
 [235 245 280 240 210 220]
 [245 255 290 205 215 225]
 [255 265 300 260 220 230]
 [265 275 300 270 225 235]]
[220 210 230 240 205 210 215 220 205 225 215 230 225 210 225 235 230 205
 215 235 245 240 210 220 245 205 215 225 220 230 225 235]
32
[200 220 210 180 190 230 200 195 205 210 215 220 185 200 205 225 190 195
 195 215 230 225 200 210 225 230 205 215 210 220 205 215 225 220 230 225]


In [ ]:
sorted_data = np.sort(sales_data[4])
print(sorted_data)

# to convert it into descending order
sorted_data[::-1]
print(sorted_data[::-1])

days_total = np.sum(sales_data, axis=1)
print(days_total)
max_day_index = np.argmax(days_total)
print(max_day_index)

#column-wise means
col = np.mean(sales_data, axis=0)
print(col)

#row-wise means
row = np.mean(sales_data, axis=1)
print(row)

overall_average = np.mean(sales_data)
print(overall_average)

[200 210 215 225 230 265]
[265 230 225 215 210 200]
[1250 1330 1285 1285 1345 1385 1430 1435 1530 1570]
9
[228.5 240.5 275.  225.  203.  212.5]
[208.33333333 221.66666667 214.16666667 214.16666667 224.16666667
 230.83333333 238.33333333 239.16666667 255.         261.66666667]
230.75


In [ ]:
def highlight_outliers(arr):
     below = np.mean(arr.flatten()) - (2*np.std(arr.flatten()))
     above = np.mean(arr.flatten()) + (2*np.std(arr.flatten()))
     print(above)
     print(below)
     return arr[(arr<below) | (arr>above)]


highlight_outliers(sales_data)

289.9340350094517
171.56596499054834


array([290, 300, 300])

**Task Day-02-03**

**Mini Project:**
You are required to implement a in Python using Object-Oriented
Programming (OOP) principles.

Student Report Card Generator

Problem:
Write a Python program using a class called to manage a student's academic
performance. The program must:

**Part A — Class Implementation**
1. Create a class Student with the following:
• Attributes: name , age , and grades (a dictionary to store subject → marks)
• Method add_grade(subject, marks) to add a subject and its marks to the dictionary

• Method calculate_average() to compute the average of all subject marks

• Method assign_grade() to return the letter grade based on average:

Average Score Grade

80 and above A

60 to 79 B

40 to 59 C

Below 40 F

2. Add a method that applies a lambda function to increase all subject
marks (max 100).

apply_bonus(lambda_func)

3. Add methods to:

• generate_report() to return a formatted report string.

• save_to_file(filename) to save the report to a .txt file.

• A @staticmethod load_from_file(filename) to read and display the report.

**Part B — Program Flow**
1. Prompt the user to enter the student’s name and age .
2. Take marks input for 3 subjects: Math, English, and Urdu.
3. Ask the user:
Do you want to apply 5 bonus marks to all subjects? (yes/no)
If yes, use a lambda function to apply the bonus

**Part C — File Handling & Output**
1. Save the report to a file named <name>_report.txt
2. Read from the file and display the report in the console.

Example Output:

Enter student's name: Ali Enter age: 20 Enter marks for Math: 90 Enter marks for English:
83 Enter marks for Urdu: 95 Add 5 bonus marks to all subjects? yes Report saved successfu
lly! Report Card: Name: Ali Age: 20 Math: 95 English: 88 Urdu: 100 Average: 94.33 Grade:
A

In [ ]:
class Student:
    def __init__(self,name,age):
        self.name = name
        self.age = age
        self.grades = {}

    def add_grade(self, subject, marks):
        self.grades[subject] = marks

    def calculate_average(self):
        if not self.grades:
            return 0
        total = sum(self.grades.values())
        average = total / len(self.grades)
        return average

    def assign_grade(self):
        avg = self.calculate_average()
        if avg >= 80:
            return 'A'
        elif avg >= 60:
            return 'B'
        elif avg >= 40:
            return 'C'
        else:
            return 'F'

    def apply_bonus(self, lambda_func):
        # Apply the lambda to each mark and cap it at 100
        self.grades = {subject: min(lambda_func(marks), 100)
                       for subject, marks in self.grades.items()}

    def generate_report(self):
        report = f"--- Student Report ---\n"
        report += f"Name: {self.name}\nAge: {self.age}\n"
        report += "\nSubject Marks:\n"
        for subject, marks in self.grades.items():
            report += f"  {subject}: {marks}\n"
        report += f"\nAverage Marks: {self.calculate_average():.2f}\n"
        report += f"Grade: {self.assign_grade()}\n"
        return report

    def save_to_file(self, filename):
        with open(filename, 'w') as file:
            file.write(self.generate_report())

    @staticmethod
    def load_from_file(filename):
        try:
            with open(filename, 'r') as file:
                content = file.read()
                print(content)
        except FileNotFoundError:
            print(f"File '{filename}' not found")



name = input("Enter student name: ")
age = int(input("Enter student age: "))

student = Student(name, age)

# Input marks for 3 subjects
for subject in ["Math", "English", "Urdu"]:
    marks = float(input(f"Enter marks for {subject}: "))
    student.add_grade(subject, marks)

# Ask for bonus
apply = input("Do you want to apply 5 bonus marks to all subjects? (yes/no): ").lower()
if apply == "yes":
    student.apply_bonus(lambda m: m + 5)

# -----------------------------
# Part C: File Handling
# -----------------------------
filename = f"{student.name}_report.txt"
student.save_to_file(filename)
print("\n--- Report Saved ---\n")

print("--- Report Loaded From File ---\n")
Student.load_from_file(filename)

Enter student name: Kinza
Enter student age: 20
Enter marks for Math: 95
Enter marks for English: 84
Enter marks for Urdu: 90
Do you want to apply 5 bonus marks to all subjects? (yes/no): Yes

--- Report Saved ---

--- Report Loaded From File ---

--- Student Report ---
Name: Kinza
Age: 20

Subject Marks:
  Math: 100.0
  English: 89.0
  Urdu: 95.0

Average Marks: 94.67
Grade: A



**More Advanced Operations from Numpy**

**Solving linear Equations:**

**Ax= B**

Where:

A is a square matrix (e.g., 2x2, 3x3).

B is a vector or matrix of known values.

The goal is to find the vector x that satisfies the equation.

✅ Example:
Let's say you have the system:

2x+3y=8

3x+4y=11

In [ ]:
import numpy as np
A = np.array([[2, 3], [3, 4]])
B = np.array([8, 11])

x = np.linalg.solve(A, B)
print(x)

[1. 2.]


**Determinant of a Matrix:**

np.linalg.det(matrix)

In [ ]:
A = np.array([[1, 2], [3, 4]])
det = np.linalg.det(A)
print(det)

-2.0000000000000004


**Inverse of a Matrix:**

np.linalg.inv(matrix)

In [ ]:
A = np.array([[1, 2], [3, 4]])
inv_A = np.linalg.inv(A)
print(inv_A)

identity = np.dot(A, inv_A)
print(identity)

#This is due to floating-point precision — computers can't represent decimal numbers with perfect accuracy, so you get very tiny numbers instead of exact zeros.
#For example:
#8.8817842e-16 is scientific notation for:

#8.88×10−16 = 0.00000000000000088817842
#This is extremely close to 0, so we can treat it as 0 in mathematical terms.

[[-2.   1. ]
 [ 1.5 -0.5]]
[[1.0000000e+00 0.0000000e+00]
 [8.8817842e-16 1.0000000e+00]]


**Random Number Generation:**

**✅ np.random.seed()**
This function is used to set the random seed so that random numbers are reproducible — i.e., every time you run the code, you get the same random numbers.

🧠 Why use it?
Without it, random numbers will change every time you run the code.

**✅ np.random.rand()**
This function generates random floats between 0 and 1 from a uniform distribution.   

np.random.rand()      # Single float between 0 and 1

np.random.rand(3)     # 1D array of 3 random floats

np.random.rand(2, 3)  # 2x3 array of random floats


**✅np.random.randint(low, high, size)**

📌 Parameters:

low: Start (inclusive) — minimum value allowed.

high: End (exclusive) — max value not included.

size: How many random numbers or what shape (optional).

In [ ]:
import numpy as np
print(np.random.rand())
print(np.random.rand(2,3))
print(np.random.rand(4))

np.random.seed(42)
print(np.random.rand())
print(np.random.rand(2,3))
print(np.random.rand(4))

print(np.random.randint(0,100,5))
print(np.random.randint(0,100,size=(1,2)))
print(np.random.randint(0,100))

0.18340450985343382
[[0.30424224 0.52475643 0.43194502]
 [0.29122914 0.61185289 0.13949386]]
[0.29214465 0.36636184 0.45606998 0.78517596]
0.3745401188473625
[[0.95071431 0.73199394 0.59865848]
 [0.15601864 0.15599452 0.05808361]]
[0.86617615 0.60111501 0.70807258 0.02058449]
[ 1 87 29 37  1]
[[63 59]]
20


**Task Day-05**   
Task: Build a NumPy-Based Mini ML Engine for Linear Regression with
Synthetic Data.   

Objective:
Create a small Python app using NumPy that generates synthetic data, performs linear
regression using matrix operations, and supports saving/loading the data and model weights.

**Key Requirements (Mapped to Topics)**

**1) Linear Algebra Operations (ML-Relevant)**

• Use element-wise multiplication and dot product to compute predictions.

• Apply transpose and inverse for solving the Normal Equation:

2x+y = 3
x+3y = 4

• Use np.sum(np.diag(...)) to compute sum of diagonal elements in matrix
covariance.

• Use np.linalg.solve as an alternative method for solving linear systems.

**2) Random Number Generation**

• Use np.random.seed() for reproducibility.

• Use np.random.random() or np.random.randint() to generate synthetic feature data.

• Use np.random.choice() to randomly select samples for testing/training splits.

**3) NumPy I/O Operations**

• Save and load model weights using np.save() and np.load().

• Save the dataset to .txt format using np.savetxt() and load it back using
np.loadtxt().

• Use np.savez() to store multiple arrays in a single file, e.g., features, targets, and
weights.

**4) Arithmetic Operations and Array Slicing**

Apply vectorized arithmetic operations to perform statistical analysis.
Use slicing to separate train/test sets or to manipulate arrays during operations.

**Bonus (Optional)**

• Allow user input to choose between .npy, .txt, or .npz formats for saving data.

In [1]:
import numpy as np
# 1. Element-wise multiplication & dot product
# Example vectors
features = np.array([2, 3])
weights = np.array([0.5, 1.5])

# Element-wise multiplication
elementwise_product = features * weights
print("Element-wise multiplication:", elementwise_product)

# Dot product (prediction in ML)
dot_product = np.dot(features, weights)
print("Dot product (prediction):", dot_product)

# 2. Solving Linear System using Transpose & Inverse (Normal Equation)
# System of equations:
# 2x + y = 3
# x + 3y = 4

# Matrix form: AX = B
A = np.array([[2, 1],
              [1, 3]])
B = np.array([3, 4])

# Using inverse: X = A⁻¹B
A_inv = np.linalg.inv(A)
X = np.dot(A_inv, B)
print("Solution using inverse:\n", X)

# 3. Sum of Diagonal Elements (Trace of Covariance Matrix)
# Example matrix (covariance-like)
matrix = np.array([[4, 2, 1],
                   [2, 5, 3],
                   [1, 3, 6]])

# Diagonal sum (trace)
trace = np.sum(np.diag(matrix))
print("Sum of diagonal elements (trace):", trace)

# 4. Solving Linear System using np.linalg.solve
# Same system: AX = B
X_alt = np.linalg.solve(A, B)
print("Solution using np.linalg.solve:\n", X_alt)


Element-wise multiplication: [1.  4.5]
Dot product (prediction): 5.5
Solution using inverse:
 [1. 1.]
Sum of diagonal elements (trace): 15
Solution using np.linalg.solve:
 [1. 1.]


In [14]:
# Random number generation

#seed for reproducibility
np.random.seed(42)

# Random data with 10 rows and 3 columns/features
synthetic_data = np.random.randint(0,100,size=(10,3))
print(synthetic_data)

# Selecting data randomly for training
indices =np.arange(10)
print(indices)
#lets assume we have 10 datasets
#Randomly select 70% (7 out of 10) for training
train_data_indices = np.random.choice(indices,size = 7,replace=False)
print(train_data_indices)
train_data = synthetic_data[train_data_indices]
print(train_data)
#Select rest of the data for testing
test_indices = np.setdiff1d(indices,train_data_indices )
test_data = synthetic_data[test_indices]
print(test_data)


[[51 92 14]
 [71 60 20]
 [82 86 74]
 [74 87 99]
 [23  2 21]
 [52  1 87]
 [29 37  1]
 [63 59 20]
 [32 75 57]
 [21 88 48]]
[0 1 2 3 4 5 6 7 8 9]
[1 5 4 8 0 7 6]
[[71 60 20]
 [52  1 87]
 [23  2 21]
 [32 75 57]
 [51 92 14]
 [63 59 20]
 [29 37  1]]
[[82 86 74]
 [74 87 99]
 [21 88 48]]


In [17]:
#Numpy I/O Operations
features = np.random.randint(0,100,size=(5,3))
targets = np.random.randint(0,2,size=(5,))
weights = np.random.random(size=(3,))

print("Features:\n", features)
print("Targets:\n", targets)
print("Weights:\n", weights)

np.save('model_weights.npy', weights)              # save weights
loaded_weights = np.load('model_weights.npy')      # load weights
print("\nLoaded Weights:\n", loaded_weights)

np.savetxt('features.txt', features, fmt='%d')      # save as text
loaded_features = np.loadtxt('features.txt', dtype=int)
print("\nLoaded Features from TXT:\n", loaded_features)

np.savez('model_data.npz', features=features, targets=targets, weights=weights)
data = np.load('model_data.npz')
print("\nLoaded from NPZ:")
print("Features:\n", data['features'])
print("Targets:\n", data['targets'])
print("Weights:\n", data['weights'])


Features:
 [[39 84 79]
 [81 52 23]
 [25 88 59]
 [40 28 14]
 [44 64 88]]
Targets:
 [0 0 1 0 1]
Weights:
 [0.29348817 0.01407982 0.1988424 ]

Loaded Weights:
 [0.29348817 0.01407982 0.1988424 ]

Loaded Features from TXT:
 [[39 84 79]
 [81 52 23]
 [25 88 59]
 [40 28 14]
 [44 64 88]]

Loaded from NPZ:
Features:
 [[39 84 79]
 [81 52 23]
 [25 88 59]
 [40 28 14]
 [44 64 88]]
Targets:
 [0 0 1 0 1]
Weights:
 [0.29348817 0.01407982 0.1988424 ]
